### PR TITLE
Nacos unsub support and RegistryDirectory destroy

### PR DESCRIPTION
**Pull request by owen200008**
1. nacos support doUnsubscribe
2. MockClusterInvoker destroy need destroy directory

#8895 second question， notify and recreate the destroy ReferenceConfig connect

### CODE REVIEW

**Comment by plusmancn** [comment]
> #8895 second question， notify and recreate the destroy ReferenceConfig connect

Try to use `Fixes #8895` in your context.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/MockClusterInvoker.java [hunk 1]
```diff
@@ -75,6 +75,9 @@ public boolean isAvailable() {
 
     @Override
     public void destroy() {
+        //#8895 second question directory must destroy
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/wrapper/MockClusterInvoker.java:78]: please remove this issue reference, just leave some description is enough

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/DynamicDirectory.java [hunk 2]
```diff
@@ -238,7 +238,8 @@ public void destroy() {
         // unsubscribe.
         try {
             if (getSubscribeConsumerurl() != null && registry != null && registry.isAvailable()) {
-                registry.unsubscribe(getSubscribeConsumerurl(), this);
+                //#8895 second question overwrite by child
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/DynamicDirectory.java:241]: please remove this issue reference, just leave some description is enough

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java [hunk 3]
```diff
@@ -495,26 +512,45 @@ private void filterServiceNames(Set<String> serviceNames, URL url) {
 
     private void subscribeEventListener(String serviceName, final URL url, final NotifyListener listener)
             throws NacosException {
-        EventListener eventListener = event -> {
-            if (event instanceof NamingEvent) {
-                NamingEvent e = (NamingEvent) event;
-                List<Instance> instances = e.getInstances();
+        ConcurrentMap<NotifyListener, EventListener> listeners = nacosListeners.computeIfAbsent(url, k -> new ConcurrentHashMap<>());
+        EventListener nacosListener = listeners.computeIfAbsent(listener, k -> {
```
Comment (wuwen5) [inline dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java:516]: mark: This fixes a memory leak in Dubbo admin √

Comment (wuwen5) [reply dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosRegistry.java:516]: 发现这里解决了一个内存泄露问题👍，之前```EventListener```会被```scheduleServiceNamesLookup```无限制存储到 ```om.alibaba.nacos.client.naming.event.InstancesChangeNotifier#listenerMap```中，导致内存溢出, 这里标记下，后续有人dubbo-admin中遇到可以关注下

**Comment by wuwen5** [comment]
NacosRegistry 存在一个严重的内存泄露问题，在dubbo admin中出现，经过分析发现被这个pr解决了，建议2.7.x尽早发布一个release版本.

```
private void subscribeEventListener(String serviceName, final URL url, final NotifyListener listener)
            throws NacosException {
        EventListener eventListener = event -> {
            if (event instanceof NamingEvent) {
                NamingEvent e = (NamingEvent) event;
                List<Instance> instances = e.getInstances();


                if (isServiceNamesWithCompatibleMode(url)) {
                    /**
                     * Get all instances with corresponding serviceNames to avoid instance overwrite and but with empty instance mentioned
                     * in https://github.com/apache/dubbo/issues/5885 and https://github.com/apache/dubbo/issues/5899
                     */
                    NacosInstanceManageUtil.initOrRefreshServiceInstanceList(serviceName, instances);
                    instances = NacosInstanceManageUtil.getAllCorrespondingServiceInstanceList(serviceName);
                }

                notifySubscriber(url, listener, instances);
            }
        };
        namingService.subscribe(serviceName,
                getUrl().getParameter(GROUP_KEY, Constants.DEFAULT_GROUP),
                eventListener);
    }
```
```EventListener eventListener```实例会无限制存储到 nacos ```InstancesChangeNotifier#listenerMap```的ConcurrentHashSet中

```
public class InstancesChangeNotifier extends Subscriber<InstancesChangeEvent> {
    
    private final Map<String, ConcurrentHashSet<EventListener>> listenerMap = new ConcurrentHashMap<String, ConcurrentHashSet<EventListener>>();
    
    private final Object lock = new Object();
    
    /**
     * register listener.
     *
     * @param serviceName combineServiceName, such as 'xxx@@xxx'
     * @param clusters    clusters, concat by ','. such as 'xxx,yyy'
     * @param listener    custom listener
     */
    public void registerListener(String serviceName, String clusters, EventListener listener) {
        String key = ServiceInfo.getKey(serviceName, clusters);
        ConcurrentHashSet<EventListener> eventListeners = listenerMap.get(key);
        if (eventListeners == null) {
            synchronized (lock) {
                eventListeners = listenerMap.get(key);
                if (eventListeners == null) {
                    eventListeners = new ConcurrentHashSet<EventListener>();
                    listenerMap.put(key, eventListeners);
                }
            }
        }
        eventListeners.add(listener);
    }
...
}
```

E1E5323A-BF00-4F1D-96D3-E2A07374244B

884E8D77-67AB-417E-ABC4-7873057DDC74

**Comment by wuwen5** [comment]
@AlbumenJ  Online problems urgently require a hotfix version

**Pull request merged**